lnrpc/invoicesrpc/invoices_server.go

```diff
@@ -101,7 +104,55 @@ var _ InvoicesServer = (*Server)(nil)
 // this method. If the macaroons we need aren't found in the filepath, then
 // we'll create them on start up. If we're unable to locate, or create the
 // macaroons we need, then we'll return with an error.
-func New(cfg *Config) (*Server, lnrpc.MacaroonPerms, error) {
+func New() (*Server, lnrpc.MacaroonPerms, error) {
+	server := &Server{
+		cfg:  &Config{},
+		quit: make(chan struct{}, 1),
+	}
+
+	return server, macPermissions, nil
+}
+
+// Start launches any helper goroutines required for the Server to function.
+//
+// NOTE: This is part of the lnrpc.SubServer interface.
+func (s *Server) Start() error {
+	return nil
+}
+
+// Stop signals any active goroutines for a graceful closure.
+//
+// NOTE: This is part of the lnrpc.SubServer interface.
+func (s *Server) Stop() error {
+	close(s.quit)
+
+	return nil
+}
+
+// InjectDependencies populates the sub-server's dependencies. If the
+// finalizeDependencies boolean is true, then the sub-server will finalize its
+// dependencies and return an error if any required dependencies are missing.
+//
+// NOTE: This is part of the lnrpc.SubServer interface.
+func (s *Server) InjectDependencies(
```

Comment:
What's the rationale re why this can't just be part of the `New` config?

Reply:
So the core issue is that we need the keys, in order to populate some of the current dependencies for the sub-servers, i.e. in order to run `PopulateDependencies` https://github.com/lightningnetwork/lnd/blob/e8c5e7d5ce4b79117ef9ab497745fea3a1bffb36/rpcserver.go#L757 That therefore requires that the remote signer is connected before that line is executed.

The issue with that is though, that the current `lnd` code base won't allow any incomming RPC requests before the `addDeps` function has finished executing (which executes `PopulateDependencies`). After the `addDeps` has been executed, we proceed to set the RPC state to Active.

We therefore need to allow one specific `RPC` request (namely the one where the remote signer connects), before that line is executed.

We also aimed to maintain the functionality in `lnd` that ensures that we always create all RPC sub-servers in same code block, to not make that more complicated. But as the previous RPC sub-servers' `New` function required that the dependencies already existed, I needed to separate adding of the dependencies from the `New` function. That's why this `InjectDependencies` now exists, which is executed for the sub-servers, after the remote signer has connected.